smm/lotus_adapter_test.go

```diff
@@ -0,0 +1,386 @@
+package smm
+
+import (
+    "bytes"
+    "context"
+    "crypto/rand"
+    "github.com/filecoin-project/lotus/api/test"
+    "github.com/filecoin-project/lotus/chain/address"
+    "github.com/filecoin-project/lotus/miner"
+    "github.com/filecoin-project/lotus/node"
+    "github.com/filecoin-project/lotus/node/modules"
+    modtest "github.com/filecoin-project/lotus/node/modules/testing"
+    "github.com/filecoin-project/lotus/node/repo"
+    "github.com/ipfs/go-cid"
+    "github.com/libp2p/go-libp2p-core/crypto"
+    "github.com/libp2p/go-libp2p-core/peer"
+    mocknet "github.com/libp2p/go-libp2p/p2p/net/mock"
+    "github.com/multiformats/go-multihash"
+    "testing"
+)
+
+type testListener struct {
+    Called bool
+}
+
+func (t testListener) OnChainStateChanged(*StateChange) {
+    t.Called = true
+}
+
+func create(ctx context.Context) (*test.TestNode, error) {
+    var testNode test.TestNode
+    mn := mocknet.New(ctx)
+
+    pk, _, err := crypto.GenerateEd25519Key(rand.Reader)
+    if err != nil {
+        return nil, err
+    }
+    minerPid, err := peer.IDFromPrivateKey(pk)
+    var genbuf bytes.Buffer
+    mineBlock := make(chan struct{})
+    _, err = node.New(ctx,
+        node.FullAPI(&testNode.FullNode),
+        node.Online(),
+        node.Repo(repo.NewMemory(nil)),
+        node.MockHost(mn),
+        node.Test(),
+        node.Override(new(*miner.Miner), miner.NewTestMiner(mineBlock)),
+        node.Override(new(modules.Genesis), modtest.MakeGenesisMem(&genbuf, minerPid)),
+    )
+    if err != nil {
+        return nil, err
+    }
+    testNode.MineOne = func(ctx context.Context) error {
+        select {
+        case mineBlock <- struct{}{}:
+            return nil
+        case <-ctx.Done():
+            return ctx.Err()
+        }
+    }
+    err = mn.LinkAll()
+    if err != nil {
+        return nil, err
+    }
+    return &testNode, nil
+}
+
+func Test_NodeCreation(t *testing.T) {
+    ctx := context.Background()
+    fullnode, err := create(ctx)
+    if err != nil {
+        t.Fatal(err)
+    }
+    actor, err := fullnode.WalletDefaultAddress(ctx)
+    if err != nil {
+        t.Fatal(err)
+    }
+    miner, err := address.NewFromString("t0101")
+    if err != nil {
+        t.Fatal(err)
+    }
+    listener := testListener{false}
+    node, initialState, err := NewNode(ctx, fullnode, Address(actor.String()), Address(miner.String()), listener)
```

Comment:
This name `node` is becoming really confusing from the outside, where it's definitely not the node. I think we should probably rename to `NewStorageMinerAdapter` or similar.